README
This directory contains the 0.20.0 release of Timeline.

Timeline is a cross-platform application for displaying and navigating events
on a timeline.

Timeline is free software, distributed under the GNU General Public License,
version 3. For more information, see the file COPYING.

Icons created by us are licensed under the Creative Commons Attribution-Share
Alike 3.0 Unported License. Some icons were also taken from the Tango Desktop
Project. For more information, see file icons/COPYING.

Translations are licensed under the modern, simplified BSD license and are
managed on Launchpad: http://translations.launchpad.net/thetimelineproj.

To learn what is new in this version of Timeline, see CHANGES.

For installation instructions, see INSTALL.

For information about working with the source code, see HACKING.

For project status, see STATUS.

For more information about Timeline, visit its website at
http://thetimelineproj.sourceforge.net/ or send an email to the user mailing
list at <thetimelineproj-user@lists.sourceforge.net>.


CHANGES
Timeline 0.20.0, released on 30 March 2013
==========================================

New features, enhancements:

  * Added 'Save As' feature
  * Strategy for allowing multiple users to use the same Timeline file.
  * The timeline view regains focus when the event editor is closed.
  * Enter-key works in date and time fields of the event editor
  * Some help texts updated
  * New version of icalender to cope with years before 1900
  * TimelineComponent can explicitly clear the drawing area
  
Bug fixes:

  * Fixed problem with Event texts starting with '('- or '['-character
  * Delete event by context menu now works


Timeline 0.19.0, released on 30 December 2012
=============================================

New features, enhancements:

  * Possibility to define URL on events and execute "Goto URL" to open web browser.
  * Implemented 'fit week' navigation function.
  * Help text added, to describe vertical movement of events.
  
Bug fixes:

  * Build script generates zip file with only LF as line endings in files
  * Year 0 removed from timeline display when using extended date range

Timeline 0.18.0, released on 30 September 2012
==============================================

New features, enhancements:

  * Zooming with scroll wheel zooms at cursor position instead of center.

Bug fixes:

  * Adding multiple events without closing event dialog, works again.
  * Alert time comparision problem solved
  * Fixed problem with ends-today property
  * Fit millennium now works close to edges
  * Fit century now works close to edges

Timeline 0.17.0, released on 15 June 2012
=========================================

This is a new feature release.

New features, enhancements:

  * Possibilty to define alerts on events.
  * Non-period events can be added to container events

Bug fixes:

  * No Error when fitting month, december, when using extended timetype.

Timeline 0.16.0, released on 31 January 2012
============================================

This is a new feature release.

New features, enhancements:

  * Events can be grouped in containers

Bug fixes:

  * Timeline files with non-English names can be opened
  * Creating new locked events does not raise exception

Timeline 0.15.0, released on 30 October 2011
============================================

This is a new feature release.

New features, enhancements:

  * Custom font color for categories
  * Measure distance between events
  * Only break text in balloon if needed to keep balloon on screen

Bug fixes:

  * SVG export can now handle text with non-english characters
  * Long category names are now visible in category editor
  * Timeline repaints after editing category color
  * No year of out range exception in event dialog

Timeline 0.14.0, released on 30 July 2011
=========================================

This is a new feature release.

New features, enhancements:

  * Move all selected events
  * Mark event period as fuzzy and edges will change to triangles
  * Mark event period as locked and edges will be curved and the event can not
    be moved or resized
  * Mark event as ending today and its period will be updated to end today
  * Experimental support for inertial scrolling (can be enabled in preferences)
  * Shows status text when zooming

Bug fixes:

  * Not possible to select too large period when zooming with shift+drag
  * Prevent exception (in cases when year was out of range) when scrolling with
    page up/down
  * Show user friendly message when creating event with too long period
  * Display error message in status bar if period is too long when resizing
    event
  * No time exception when exporting to SVG
  * No exception when using extended date range and exporting to SVG

Timeline 0.13.0, released on 30 April 2011
==========================================

This is a new feature release.

New features, enhancements:

  * Events can be moved up and down with Alt+Up/Down
  * Hidden event count is shown in status bar
  * Event text changes color to white if background is dark
  * Timeline can be scrolled with Alt+Left/Right
  * Edit category button added in categories editor
  * Export to SVG

Bug fixes:

  * No exception if "Fit all events" results in a period too large to display
  * No error if pressing left or right in empty categories tree control

Timeline 0.12.1, released on 30 January 2011
============================================

This is a translation update and bugfix release.

Bug fixes:

  * Menu items are correctly disabled if no timeline is open
  * Clicking calendar button when an invalid date is entered gives error
    message instead of exception
  * LANG environment variable is only set on Windows to prevent locale error at
    startup on Linux systems
  * Fit all events ignores hidden events

Timeline 0.12.0, released on 9 January 2011
===========================================

This is a new feature release.

New features, enhancements:

  * Experimental support for extended date range (before 1 AD)

Bug fixes:

  * Centuries before 10th are displayed correctly (9 instead of 90)
  * Correct translations are used on Windows

New translations:

  * Lithuanian
  * Vietnamese

Timeline 0.11.1, released on 24 October 2010
============================================

This is a translation update and bugfix release.

Bug fixes:

  * Create event through menu does not raise exception
  * Time removed when saving event and 'Show time' not checked

Timeline 0.11.0, released on 12 October 2010
============================================

This is a new feature release.

New features, enhancements:

  * New improved date and time entry control
  * New navigation function: fit millennium

Bug fixes:

  * Remove import of wx.lib.wordwrap that caused a crash on Ubuntu

New translations:

  * Italian
  * Turkish

Timeline 0.10.2, released on 11 June 2010
=========================================

This is a translation update and bugfix release.

Bug fixes:


specs/MemoryDB.py
# Copyright (C) 2009, 2010, 2011  Rickard Lindberg, Roger Lindberg
#
# This file is part of Timeline.
#
# Timeline is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Timeline is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Timeline.  If not, see <http://www.gnu.org/licenses/>.


from datetime import datetime
import unittest

from mock import Mock

from timelinelib.db.backends.memory import MemoryDB
from timelinelib.db.exceptions import TimelineIOError
from timelinelib.db.objects import Category
from timelinelib.db.objects import Event
from timelinelib.db.objects import TimePeriod
from timelinelib.drawing.viewproperties import ViewProperties
from timelinelib.time import PyTimeType


class MemoryDBSpec(unittest.TestCase):

    def testInitialState(self):
        db = MemoryDB()
        self.assertEquals(db.path, "")
        self.assertEquals(db.displayed_period, None)
        self.assertEquals(db.hidden_categories, [])
        self.assertEquals(db.is_read_only(), False)
        self.assertEquals(db.supported_event_data(), ["description", "icon", "alert", "hyperlink"])
        self.assertEquals(db.search(""), [])
        self.assertEquals(db.get_all_events(), [])
        self.assertEquals(db.get_first_event(), None)
        self.assertEquals(db.get_last_event(), None)
        self.assertEquals(db.get_categories(), [])

    def testLoadSaveViewProperties(self):
        # Make sure the database contains categories
        self.db.save_category(self.c1)
        self.db.save_category(self.c2)
        # Set up a view properties object that simulates having selected a
        # specific period and hidden one category
        vp = ViewProperties()
        vp.set_category_visible(self.c1, False)
        start = datetime(2010, 3, 23)
        end = datetime(2010, 3, 24)
        tp = TimePeriod(self.db.get_time_type(), start, end)
        vp.displayed_period = tp
        # Save these properties and assert that the database fields are written
        # correctly
        self.db.save_view_properties(vp)
        self.assertEquals(self.db.displayed_period, tp)
        self.assertEquals(self.db.hidden_categories, [self.c1])
        # Load view properties from db simulating that this db was just loaded
        # into memory and the view is being configured
        new_vp = ViewProperties()
        self.db.load_view_properties(new_vp)
        self.assertFalse(new_vp.category_visible(self.c1))
        self.assertTrue(new_vp.category_visible(self.c2))
        self.assertEquals(new_vp.displayed_period, self.db.displayed_period)
        # Assert virtual _save method called: 2 save categories, 1 save view
        # properties
        self.assertEquals(self.db._save.call_count, 3)

    def testSaveInvalidDisplayedPeriod(self):
        # Assign a zero-period as displayed period
        vp = ViewProperties()
        start = datetime(2010, 3, 23)
        end = datetime(2010, 3, 23)
        tp = TimePeriod(PyTimeType(), start, end)
        vp.displayed_period = tp
        # Assert error when trying to save
        self.assertRaises(TimelineIOError, self.db.save_view_properties, vp)

    def testGetSetDisplayedPeriod(self):
        tp = TimePeriod(self.db.get_time_type(), datetime(2010, 3, 23),
                        datetime(2010, 3, 24))
        self.db._set_displayed_period(tp)
        # Assert that we get back the same period
        self.assertEquals(self.db._get_displayed_period(), tp)
        # Assert that the period is correctly loaded into ViewProperties
        vp = ViewProperties()
        self.db.load_view_properties(vp)
        self.assertEquals(vp.displayed_period, tp)

    def testGetSetHiddenCategories(self):
        # Assert that we cannot include categories not in the db
        self.assertRaises(ValueError, self.db._set_hidden_categories, [self.c1])
        self.db._set_hidden_categories([])
        self.db.save_category(self.c1)
        self.db.save_category(self.c2)
        self.db._set_hidden_categories([self.c1])
        # Assert that the returned list is the same
        self.assertEquals(self.db._get_hidden_categories(), [self.c1])
        # Assert that category visibility information is correctly written to
        # ViewProperties
        vp = ViewProperties()
        self.db.load_view_properties(vp)
        self.assertFalse(vp.category_visible(self.c1))
        self.assertTrue(vp.category_visible(self.c2))

    def testSaveNewCategory(self):
        self.db.save_category(self.c1)
        self.assertTrue(self.c1.has_id())
        self.assertEqual(self.db.get_categories(), [self.c1])
        self.assertEqual(self.db_listener.call_count, 1)
        # Assert virtual _save method called: 1 save category
        self.assertEquals(self.db._save.call_count, 1)

    def testSaveExistingCategory(self):
        self.db.save_category(self.c1)
        id_before = self.c1.id
        self.c1.name = "new name"
        self.db.save_category(self.c1)
        self.assertEqual(id_before, self.c1.id)
        self.assertEqual(self.db.get_categories(), [self.c1])
        self.assertEqual(self.db_listener.call_count, 2) # 2 save
        # Assert virtual _save method called: 2 save category
        self.assertEquals(self.db._save.call_count, 2)

    def testSaveNonExistingCategory(self):
        other_db = MemoryDB()
        other_db.save_category(self.c1)
        # It has id but is not in this db
        self.assertRaises(TimelineIOError, self.db.save_category, self.c1)
        # Assert virtual _save method not called
        self.assertEquals(self.db._save.call_count, 0)

    def testSaveCategoryWithUnknownParent(self):
        self.c1.parent = self.c2
        # c2 not in db so we should get exception
        self.assertRaises(TimelineIOError, self.db.save_category, self.c1)
        # But after c2 is added everything is fine
        self.db.save_category(self.c2)
        self.db.save_category(self.c1)

    def testSaveCategoryWithParentChange(self):
        # Start with this hierarchy:
        # c1
        #   c11
        #     c111
        #   c12
        c1 = Category("c1", (255, 0, 0), None, True, parent=None)
        c11 = Category("c11", (255, 0, 0), None, True, parent=c1)
        c111 = Category("c111", (255, 0, 0), None, True, parent=c11)
        c12 = Category("c12", (255, 0, 0), None, True, parent=c1)
        self.db.save_category(c1)
        self.db.save_category(c11)
        self.db.save_category(c111)
        self.db.save_category(c12)
        # Changing c11's parent to c12 should create the following tree:
        # c1
        #   c12
        #     c11
        #       c111
        c11.parent = c12
        self.db.save_category(c11)
        self.assertEquals(c1.parent, None)
        self.assertEquals(c12.parent, c1)
        self.assertEquals(c11.parent, c12)
        self.assertEquals(c111.parent, c11)
        # Changing c11's parent to c111 should raise exception since that would
        # create a circular parent link.
        c11.parent = c111
        self.assertRaises(TimelineIOError, self.db.save_category, c11)

    def testDeleteExistingCategory(self):
        # Add two categories to the db
        self.db.save_category(self.c1)
        self.db.save_category(self.c2)
        # Make category 1 hidden
        vp = ViewProperties()
        vp.set_category_visible(self.c1, False)
        self.db.save_view_properties(vp)
        # Assert both categories in db
        categories = self.db.get_categories()
        self.assertEquals(len(categories), 2)
        self.assertTrue(self.c1 in categories)
        self.assertTrue(self.c2 in categories)
        # Remove first (by category)
        self.db.delete_category(self.c1)
        categories = self.db.get_categories()
        self.assertEquals(len(categories), 1)
        self.assertTrue(self.c2 in categories)
        self.assertFalse(self.c1.has_id())
        self.assertFalse(self.c1 in self.db.hidden_categories)
        # Remove second (by id)
        self.db.delete_category(self.c2.id)
        categories = self.db.get_categories()
        self.assertEquals(len(categories), 0)
        self.assertFalse(self.c2.has_id())
        # Check events
        self.assertEqual(self.db_listener.call_count, 4) # 2 save, 2 delete
        # Assert virtual _save method called: 2 save category, 1 save view
        # properties, 2 delete categories
        self.assertEquals(self.db._save.call_count, 5)

    def testDeleteNonExistingCategory(self):
        self.assertRaises(TimelineIOError, self.db.delete_category, self.c1)
        self.assertRaises(TimelineIOError, self.db.delete_category, 5)
        other_db = MemoryDB()
        other_db.save_category(self.c2)
        self.assertRaises(TimelineIOError, self.db.delete_category, self.c2)
        # Assert virtual _save method not called
        self.assertEquals(self.db._save.call_count, 0)

    def testDeleteCategoryWithParent(self):
        # Create hierarchy:
        # c1
        #   c11
        #   c12
        #     c121
        c1 = Category("c1", (255, 0, 0), None, True, parent=None)
        c11 = Category("c11", (255, 0, 0), None, True, parent=c1)
        c12 = Category("c12", (255, 0, 0), None, True, parent=c1)
        c121 = Category("c121", (255, 0, 0), None, True, parent=c12)
        self.db.save_category(c1)
        self.db.save_category(c11)
        self.db.save_category(c12)
        self.db.save_category(c121)
        # Delete c12 should cause c121 to get c1 as parent
        self.db.delete_category(c12)
        self.assertEquals(c121.parent, c1)
        # Delete c1 should cause c11, and c121 to be parentless
        self.db.delete_category(c1)
        self.assertEquals(c11.parent, None)
        self.assertEquals(c121.parent, None)

    def testDelteCategoryWithEvent(self):
        # Create hierarchy:
        # c1
        #   c11
        c1 = Category("c1", (255, 0, 0), None, True, parent=None)
        c11 = Category("c11", (255, 0, 0), None, True, parent=c1)
        self.db.save_category(c1)
        self.db.save_category(c11)
        # Create event belonging to c11
        self.e1.category = c11
        self.db.save_event(self.e1)
        # Delete c11 should cause e1 to get c1 as category
        self.db.delete_category(c11)
        self.assertEquals(self.e1.category, c1)
        # Delete c1 should cause e1 to have no category
        self.db.delete_category(c1)
        self.assertEquals(self.e1.category, None)

    def testSaveEventUnknownCategory(self):
        # A new
        self.e1.category = self.c1
        self.assertRaises(TimelineIOError, self.db.save_event, self.e1)
        # An existing
        self.db.save_event(self.e2)
        self.e2.category = self.c1
        self.assertRaises(TimelineIOError, self.db.save_event, self.e2)
        # Assert virtual _save method not called
        self.assertEquals(self.db._save.call_count, 1)

    def testSaveNewEvent(self):
        self.db.save_event(self.e1)
        tp = TimePeriod(self.db.get_time_type(), datetime(2010, 2, 12),
                        datetime(2010, 2, 14))
        self.assertTrue(self.e1.has_id())
        self.assertEqual(self.db.get_events(tp), [self.e1])
        self.assertEqual(self.db.get_all_events(), [self.e1])
        self.assertEqual(self.db_listener.call_count, 1) # 1 save
        # Assert virtual _save method called: 1 save event
        self.assertEquals(self.db._save.call_count, 1)

    def testSaveExistingEvent(self):
        self.db.save_event(self.e1)
        id_before = self.e1.id
        self.e1.text = "new text"
        self.db.save_event(self.e1)
        tp = TimePeriod(self.db.get_time_type(), datetime(2010, 2, 12),
                        datetime(2010, 2, 14))
        self.assertEqual(id_before, self.e1.id)
        self.assertEqual(self.db.get_events(tp), [self.e1])
        self.assertEqual(self.db.get_all_events(), [self.e1])
        self.assertEqual(self.db_listener.call_count, 2) # 1 save
        # Assert virtual _save method called: 2 save event
        self.assertEquals(self.db._save.call_count, 2)

    def testSaveNonExistingEvent(self):
        other_db = MemoryDB()
        other_db.save_event(self.e1)
        # It has id but is not in this db
        self.assertRaises(TimelineIOError, self.db.save_event, self.e1)
        # Assert virtual _save method not called
        self.assertEquals(self.db._save.call_count, 0)

    def testDeleteExistingEvent(self):
        tp = TimePeriod(self.db.get_time_type(), datetime(2010, 2, 12),
                        datetime(2010, 2, 15))
        self.db.save_event(self.e1)
        self.db.save_event(self.e2)
        # Assert both in db
        self.assertEquals(len(self.db.get_events(tp)), 2)
        self.assertTrue(self.e1 in self.db.get_events(tp))
        self.assertTrue(self.e2 in self.db.get_events(tp))
        # Delete first (by event)
        self.db.delete_event(self.e1)
        self.assertFalse(self.e1.has_id())
        self.assertEquals(len(self.db.get_events(tp)), 1)
        self.assertTrue(self.e2 in self.db.get_events(tp))
        # Delete second (by id)
        self.db.delete_event(self.e2.id)
        self.assertFalse(self.e2.has_id())
        self.assertEquals(len(self.db.get_events(tp)), 0)
        # Check events
        self.assertEqual(self.db_listener.call_count, 4) # 2 save, 2 delete
        # Assert virtual _save method called: 2 save event, 2 delete event
        self.assertEquals(self.db._save.call_count, 4)

    def testDeleteNonExistingEvent(self):
        self.assertRaises(TimelineIOError, self.db.delete_event, self.e1)
        self.assertRaises(TimelineIOError, self.db.delete_event, 5)
        other_db = MemoryDB()
        other_db.save_event(self.e2)
        self.assertRaises(TimelineIOError, self.db.delete_event, self.c2)
        # Assert virtual _save method not called
        self.assertEquals(self.db._save.call_count, 0)

    def testDisableEnableSave(self):
        self.db.save_category(self.c1)
        # Assert virtual _save method called: save enabled by default
        self.assertEquals(self.db._save.call_count, 1)
        self.db.disable_save()
        self.db.save_category(self.c1)
        self.assertEquals(self.db._save.call_count, 1) # still 1
        self.db.enable_save()
        self.assertEquals(self.db._save.call_count, 2)
        # Now do the same thing but tell enable not to call save
        self.db.disable_save()
        self.db.save_category(self.c1)
        self.db.enable_save(False)
        self.assertEquals(self.db._save.call_count, 2)
        # Enabling when enabled should not have any effect
        self.db.enable_save()
        self.assertEquals(self.db._save.call_count, 2)

    def testMoveEventForward(self):
        self.db.save_event(self.e1)
        self.db.save_event(self.e2)
        self.db.save_event(self.e3)
        self.db.place_event_after_event(self.e1, self.e2)
        self.assertTrue(self.db.events[0] == self.e2 )
        self.assertTrue(self.db.events[1] == self.e1 )
        self.assertTrue(self.db.events[2] == self.e3 )

    def testMoveEventToEnd(self):
        self.db.save_event(self.e1)
        self.db.save_event(self.e2)
        self.db.save_event(self.e3)
        self.db.place_event_after_event(self.e1, self.e3)
        self.assertTrue(self.db.events[0] == self.e2 )
        self.assertTrue(self.db.events[1] == self.e3 )
        self.assertTrue(self.db.events[2] == self.e1 )

    def testMoveEventBackward(self):
        self.db.save_event(self.e1)
        self.db.save_event(self.e2)
        self.db.save_event(self.e3)
        self.db.place_event_before_event(self.e2, self.e1)
        self.assertTrue(self.db.events[0] == self.e2 )
        self.assertTrue(self.db.events[1] == self.e1 )
        self.assertTrue(self.db.events[2] == self.e3 )

    def testMoveEventToBeginning(self):
        self.db.save_event(self.e1)
        self.db.save_event(self.e2)
        self.db.save_event(self.e3)
        self.db.place_event_before_event(self.e3, self.e1)
        self.assertTrue(self.db.events[0] == self.e3 )
        self.assertTrue(self.db.events[1] == self.e1 )
        self.assertTrue(self.db.events[2] == self.e2 )

    def testMoveEventToOriginalPlace(self):
        self.db.save_event(self.e1)
        self.db.save_event(self.e2)
        self.db.save_event(self.e3)
        self.db.place_event_before_event(self.e2, self.e2)
        self.assertTrue(self.db.events[0] == self.e1 )
        self.assertTrue(self.db.events[1] == self.e2 )
        self.assertTrue(self.db.events[2] == self.e3 )

    def testEventShouldNotBeFuzzyByDefault(self):
        self.assertFalse(self.e1.fuzzy)

    def testEventShouldNotBeLockedByDefault(self):
        self.assertFalse(self.e1.locked)

    def setUp(self):
        self.db = MemoryDB()
        self.db._save = Mock()
        self.db_listener = Mock()
        self.c1 = Category("work", (255, 0, 0), None, True)
        self.c2 = Category("private", (0, 255, 0), None, True)
        self.e1 = Event(self.db.get_time_type(), datetime(2010, 2, 13), datetime(2010, 2, 13),
                        "holiday")
        self.e2 = Event(self.db.get_time_type(), datetime(2010, 2, 14), datetime(2010, 2, 14),
                        "work starts")
        self.e3 = Event(self.db.get_time_type(), datetime(2010, 2, 15), datetime(2010, 2, 16),
                        "period")
        self.db.register(self.db_listener)
